README.md
# NAME

Net::Groonga::HTTP - Client library for Groonga httpd.

# SYNOPSIS

    use Net::Groonga::HTTP;

    my $groonga = Net::Groonga::HTTP->new(
        end_point => 'http://127.0.0.1:10041/d/',
    );
    my $res = $groonga->status();
    use Data::Dumper; warn Dumper($res);



# DESCRIPTION

Net::Groonga::HTTP is a client library for Groonga http server.

Groonga is a fast full text search engine. Please look [http://groonga.org/](http://groonga.org/).

# CONSTRUCTOR

    Net::Groonga::HTT->new(%args);

You can create instance with following arguments:

- end\_point :Str

    API end point URL for Groonga httpd.

    Example:

        Net::Groonga::HTTP->new(end_point => 'http://127.0.0.1:10041/d/');

- ua : Furl

    Instance of Furl to access Groonga httpd.

    Example:

        Net::Groonga::HTTP->new(ua => Furl->new());

# METHODS

- `$groonga->call($function, %args)`

    Call a http server. Function name is `$function`. Pass the `%args`.

    This method returns instance of [Net::Groonga::HTTP::Response](http://search.cpan.org/perldoc?Net::Groonga::HTTP::Response).

- $groonga->load(%args)

        $groonga->load(
            table => 'Entry',
            values => \@values,
        );

    Load the data to database. This method encodes _values_ to JSON automatically, if it's arrayref.

- $groonga->select(%args)
- $groonga->status(%args)
- $groonga->select(%args)
- $groonga->delete(%args)

    You can use these methods if you are lazy.

# LICENSE

Copyright (C) tokuhirom.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

tokuhirom <tokuhirom@gmail.com>


Changes
Revision history for Perl extension Net-Groonga-HTTP

{{$NEXT}}

0.03 2013-04-25T23:18:41Z

    - Added column_create, dump.
    - Added test cases.

0.02 2013-04-25T08:05:35Z

    - Documentation enhancements.

0.01 2013-04-25T07:38:23Z

    - original version



lib/Net/Groonga/HTTP/Response.pm
package Net::Groonga::HTTP::Response;
use strict;
use warnings;
use utf8;
use JSON::XS qw(decode_json);
use Data::Page;
use Net::Groonga::Pager;

use Mouse;

has function => (
    is => 'rw',
    required => 1,
);

has args => (
    is => 'rw',
    required => 1,
);

has http_response => (
    is => 'rw',
    required => 1,
);

has data => (
    is => 'rw',
    builder => '_build_data',
);

no Mouse;

sub is_success {
    my $self = shift;
    return $self->http_response->code eq 200 && $self->return_code == 0;
}

sub _build_data {
    my $self = shift;
    return undef if $self->http_response->code ne 200;
    decode_json($self->http_response->content);
}

sub return_code {
    my $self = shift;
    Carp::croak(sprintf("%s:%s", $self->function, $self->http_response->status_line)) unless $self->data;
    $self->data->[0]->[0];
}

sub start_time {
    my $self = shift;
    Carp::croak(sprintf("%s:%s", $self->function, $self->http_response->status_line)) unless $self->data;
    $self->data->[0]->[1];
}

sub elapsed_time {
    my $self = shift;
    Carp::croak(sprintf("%s:%s", $self->function, $self->http_response->status_line)) unless $self->data;
    $self->data->[0]->[2];
}

sub result {
    my $self = shift;
    Carp::croak(sprintf("%s:%s", $self->function, $self->http_response->status_line)) unless $self->data;
    $self->data->[1];
}

sub pager {
    my $self = shift;
    return unless $self->data;
    return unless $self->return_code == 0;

    my $total_entries = $self->data->[1]->[0]->[0]->[0];
    my $limit  = $self->args->{limit}  || 10;
    my $offset = $self->args->{offset} || 0;

    Net::Groonga::Pager->new(
        limit  => $limit,
        offset => $offset,
        total_entries => $total_entries,
    );
}

sub rows {
    my $self = shift;
    return unless $self->data;

    my @rows = @{$self->data->[1]->[0]};
    my $cnt    = shift @rows;
    my $header = shift @rows;
    my @results;
    for my $row (@rows) {
        my %args;
        for (my $i=0; $i<@$header; ++$i) {
            my $key = $header->[$i]->[0];
            my $val = $row->[$i];
            $args{$key} = $val;
        }
        push @results, \%args;
    }
    return @results;
}

1;
__END__

=head1 NAME

Net::Groonga::HTTP::Response - Response object for Net::Groonga::HTTP

=head1 DESCRIPTION

This class is a response class for L<Net::Groonga::HTTP>.

=head1 BASIC METHODS

=over 4

=item $res->function() :Str

The name of executed function.

=item $res->args() : HashRef

The arguments for executed function.

=item $res->http_response() :Object

Executed HTTP response object from L<Furl>.

=back

=head1 METHODS FOR ANALYZING CONTENT-BODY

Following methods return method dies if the response is not I<200 OK>.

=over 4

=item $res->data() :Object

JSON decoded content body.

=item $res->return_code() :Int

Shorthand for C<< $res->data->[0]->[0] >>.

Groonga's return code. It's not HTTP status code.

=item $res->starttime() :Int

Shorthand for C<< $res->data->[0]->[2] >>.

=item $res->elapsed_time() :Int

Shorthand for C<< $res->data->[0]->[2] >>.

Elapsed time.

=item $res->result() :Int

Shorthand for C<< $res->data->[1] >>.

=item $res->pager() :Net::Groonga::Pager

Create pager object if it's available.

It's only useful if the function is B<select>.

The object instance of L<Net::Groonga::Pager>.

=item @rows = $res->rows()

Create list of hashrefs from JSON content.

It's only useful if the function si B<select>.

=back




lib/Net/Groonga/HTTP.pm
package Net::Groonga::HTTP;
use 5.008005;
use strict;
use warnings;

our $VERSION = "0.02";

use JSON::XS qw(encode_json);
use Furl;
use URI;
use Net::Groonga::HTTP::Response;

use Mouse;

has end_point => ( is => 'ro', required => 1 );

has ua => (
    is => 'ro',
    default => sub {
        Furl->new(
            agent => "Net::Groonga::HTTP/$VERSION",
            timeout => 3
        )
    }
);

no Mouse;

sub construct_api_url {
    my ($self, $name, %args) = @_;
    my $url = $self->end_point;
    $url =~ s!/$!!;
    my $uri = URI->new("$url/$name");
    $uri->query_form(%args);
    $uri;
}

sub call {
    my ($self, $function, %args) = @_;
    my $url = $self->construct_api_url($function, %args);
    my $res = $self->ua->get($url);
    return Net::Groonga::HTTP::Response->new(
        function      => $function,
        http_response => $res,
        args          => \%args,
    );
}

my @functions = qw(
    table_create
    column_create
    status
    select
    delete
    dump
);
for my $function (@functions) {
    no strict 'refs';
    *{__PACKAGE__ . "::${function}"} = sub {
        my ($self, %args) = @_;
        $self->call($function, %args);
    };
}

sub load {
    my ($self, %args) = @_;
    $args{values} = encode_json($args{values}) if ref $args{values};
    return $self->call('load', %args);
}

1;
__END__

=encoding utf-8

=head1 NAME

Net::Groonga::HTTP - Client library for Groonga httpd.

=head1 SYNOPSIS

    use Net::Groonga::HTTP;

    my $groonga = Net::Groonga::HTTP->new(
        end_point => 'http://127.0.0.1:10041/d/',
    );
    my $res = $groonga->status();
    use Data::Dumper; warn Dumper($res);


=head1 DESCRIPTION

Net::Groonga::HTTP is a client library for Groonga http server.

Groonga is a fast full text search engine. Please look L<http://groonga.org/>.

=head1 CONSTRUCTOR

    Net::Groonga::HTT->new(%args);

You can create instance with following arguments:

=over 4

=item end_point :Str

API end point URL for Groonga httpd.

Example:

    Net::Groonga::HTTP->new(end_point => 'http://127.0.0.1:10041/d/');

=item ua : Furl

Instance of Furl to access Groonga httpd.

Example:

    Net::Groonga::HTTP->new(ua => Furl->new());

=back

=head1 METHODS

=over 4

=item C<< $groonga->call($function, %args) >>

Call a http server. Function name is C<< $function >>. Pass the C<< %args >>.

This method returns instance of L<Net::Groonga::HTTP::Response>.

=item $groonga->load(%args)

    $groonga->load(
        table => 'Entry',
        values => \@values,
    );

Load the data to database. This method encodes I<values> to JSON automatically, if it's arrayref.

=item $groonga->select(%args)

=item $groonga->status(%args)

=item $groonga->select(%args)

=item $groonga->delete(%args)

=item $groonga->column_create(%args)

=item $groonga->dump(%args)

You can use these methods if you are lazy.

=back

=head1 LICENSE

Copyright (C) tokuhirom.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

tokuhirom E<lt>tokuhirom@gmail.comE<gt>

=cut

